pytorch_lightning/plugins/layer_sync.py

```diff
@@ -0,0 +1,98 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from abc import ABC, abstractmethod
+
+import torch
+from torch.nn import Module
+
+
+class LayerSync(ABC):
+    """Abstract base class for creating plugins that wrap layers of a model with synchronization logic for
+    multiprocessing."""
+
+    @abstractmethod
+    def apply(self, model: Module) -> Module:
+        """Override this method to apply synchronization to the layers of this model."""
+
+    @abstractmethod
+    def revert(self, model: Module) -> Module:
+        """Override this method to undo all modifications made in :meth:`apply`."""
+
+
+class NativeSyncBatchNorm(LayerSync):
+    """A plugin that wraps all batch normalization layers of a model with synchronization logic for
+    multiprocessing.
+
+    This plugin has no effect in single-device operation.
+    """
+
+    def apply(self, model: Module) -> Module:
+        """Add global batchnorm for a model spread across multiple GPUs and nodes.
+
+        Override this method to synchronize batchnorm layers between specific process groups instead
+        of the whole world.
+
+        Args:
+            model: Reference to the current LightningModule
+
+        Return:
+            LightningModule with batchnorm layers synchronized within the process groups.
+        """
+        return torch.nn.SyncBatchNorm.convert_sync_batchnorm(model)
+
+    def revert(self, model: Module) -> Module:
+        """Convert the wrapped batchnorm layers back to regular batchnorm layers.
+
+        Args:
+            model: Reference to the current LightningModule
+
+        Return:
+            LightningModule with regular batchnorm layers that will no longer sync across processes.
+        """
+        return _revert_sync_batchnorm(model)
```

Comment:
I saw that you moved them but why not move the `_revert_sync_batchnorm` directly inside of this method?

Reply:
Because we will have the apex version for sync batchnorm and will probably want to share code.

Reply:
I suggest that we move them outside when we want to share it, just in case it doesn't work out later.